diesel/src/pg/expression/functions.rs

```diff
@@ -687,3 +688,28 @@ define_sql_function! {
     #[cfg(feature = "postgres_backend")]
     fn daterange(lower: Nullable<Date>, upper: Nullable<Date>, bound: RangeBoundEnum) -> Daterange;
 }
+
+#[cfg(feature = "postgres_backend")]
+define_sql_function! {
+    /// Append an element to the end of an array.
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::array_append;
+    /// #     use diesel::sql_types::{Integer, Array};
+    /// #     let connection = &mut establish_connection();
+    /// let ints = diesel::select(array_append::<Integer, Array<_>, _, _>(vec![1, 2], 3))
+    ///     .get_result::<Vec<i32>>(connection)?;
+    /// assert_eq!(vec![1, 2, 3], ints);
+    /// #     Ok(())
+    /// # }
+    /// ```
+    fn array_append<T: SingleValue, Arr: ArrayOrNullableArray<Inner=T> + SingleValue>(a: Arr, e: T) -> Array<T>;
```

Comment:
We should have the following two properties (unless I'm mistaken on how postgresql behaves) 

1. The SQL type of the resulting expression is Nullable depending on whether Arr is Nullable (assuming postgres returns NULL if passing a NULL array) 
2. The SQL type of the resulting expression is *not* Nullable depending on whether T is Nullable (assuming that postgres would append a NULL to the array, not make the whole resulting expression NULL) 

Can you write tests that showcase that both these things work as expected? 

Thanks 😊🙏

It is fine if the 
4.

Reply:
Thanks for the review!

The pg behavior for this functions with NULL is listed bellow:
 1.  `SELECT array_append(ARRAY[1,2], 3)` returns `ARRAY[1,2,3]`
 1.  `SELECT array_append(ARRAY[1,2], NULL)` returns `ARRAY[1, 2, NULL]`
 1. `SELECT array_append(NULL, 3)` returns `ARRAY[3]`
 1. `SELECT array_append(NULL, NULL)` returns `ARRAY[NULL]`
 
 > 1. The SQL type of the resulting expression is Nullable depending on whether Arr is Nullable (assuming postgres returns NULL if passing a NULL array)
 
 If passing NULL, postgres apparently handles it as an empty array, so I can't think of a way that this function returns NULL.
 
 > 2. The SQL type of the resulting expression is not Nullable depending on whether T is Nullable (assuming that postgres would append a NULL to the array, not make the whole resulting expression NULL)
 
 I see, thanks for the clarification. Do you know how to do that? That's why I put Nullable in everything before.
 
 > Can you add tests that showcase that both these things work as expected?
 
 Sure! Should I create doc tests or normal ones?
 
 Thanks!

Reply:
Ah I thought that we may have to involve the nullability propagation helpers traits here to construct the return type, which would maybe have involved either declaring it without the macro or [adding a feature to do so in the macro](https://github.com/diesel-rs/diesel/blob/4763e7c9c968e20d7dff7e23ba027fd1de30c37d/diesel_derives/src/sql_function.rs#L124), however given what you're saying (3 and 4) it seems that we only ever want to return an array, regardless of original nullability of the array, so it seems that the current signature is correct. 😊
(Which is furtunate because that turns out to be significantly simpler.) 

It's fine as a first version if a user needs the input array to already have its elements be Nullable to be able to append potentially null values to it. That may be incomplete (although I doubt there would be many use cases), but most importantly it's correct 😊. 
What wouldn't have been fine is if it could return nulls but that wasn't expressed in the signature (1), or if by fixing (1) we accidentally made (2) wrong (by considering the arguments in the same way, in a mode where if any argument is Nullable resulting value is Nullable, which IIRC is already our default behavior for operators but apparently not for functions - what I meant is in that case we couldn't just "enable" that behavior for functions as well).

OK so overall, current implementation looks good, we can probably just add tests to make sure that we get the correct type for all 4 scenarios, with original type of the array already expressed. 

I have no strong opinion on where each test for the 4 scenarios you mentioned should live, but maybe another reviewer will. By default unless too verbose I'd probably just have them all in the one doctest.

Reply:
Thanks! I added the test as a docstring